pkg/tfbridge/detailed_diff.go

```diff
@@ -149,6 +152,37 @@ func (differ detailedDiffer) getEffectiveType(path walk.SchemaPath) shim.ValueTy
 	return tfs.Type()
 }
 
+type hashIndexMap map[int]int
+
+func (differ detailedDiffer) calculateSetHashIndexMap(path propertyPath, listVal resource.PropertyValue) hashIndexMap {
+	identities := make(hashIndexMap)
+
+	tfs, ps, err := lookupSchemas(path, differ.tfs, differ.ps)
+	if err != nil {
+		return nil
+	}
+
+	convertedVal, err := makeSingleTerraformInput(context.Background(), path.String(), listVal, tfs, ps)
+	if err != nil {
+		return nil
+	}
+
+	if convertedVal == nil {
+		return nil
+	}
+
+	convertedListVal, ok := convertedVal.([]interface{})
+	contract.Assertf(ok, "converted value should be a list")
+
+	// Calculate the identity of each element
+	for i, newElem := range convertedListVal {
+
+		hash := tfs.SetHash(newElem)
```

Comment:
This is a little sketchy since `tfs.SetHash` accepts `interface{}` and might come to bite us. There may be slight differences. I'm mostly worried about unknowns actually, I think pulumi `interface{}` representation for TF data has unknowns that may differ from TF>

There's also a few gems I found. TF has this code:

```
func (s *Set) hash(item interface{}) string {
	code := s.F(item)
	// Always return a nonnegative hashcode.
	if code < 0 {
		code = -code
	}
	return strconv.Itoa(code)
}

```

Not sure we do that. Ah we do

```
func (s v2Schema) SetHash(v interface{}) int {
	code := s.tf.ZeroValue().(*schema.Set).F(v)
	if code < 0 {
		return -code
	}
	return code
}

```

We also have no feasible SetHash implementations in PF, running a head a little bit. It's all panic or return 0.

https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pf/proto/element.go#L96

I'm not sure if you've thought about this already. 

We have code like setEncoder

https://github.com/pulumi/pulumi-terraform-bridge/blob/master/pkg/convert/set.go#L46 ..

Reply:
tftypes.Value is awkward though as PF generally has no concept of "set element maps to an int" for comparison purposes. It can compare something for equality Eq(a, b) => bool though I think inside the set.

When we will be retrofitting this code to run over PF, it well may run into trouble as we'll be passing PF shim.Schema elements into this code that will be using SDKv2 methods to convert things with SDKv2 assumptions. It is not entirely unwrarranted since I think openTofu CLI is using cty.Value to handle PF as well but PF may just be different enough.

Reply:
There is this though that may suffice:

https://github.com/hashicorp/terraform-plugin-go/blob/v0.22.1/tftypes/value.go#L211

Once you have a terraform type T, and you have Equal(T,T) bool, you can reformulate this code to be abstract and work for either. You need:

```
type T
func Equal(T, T) bool
func ToTF(PropertyValue) T
```

You get:

```
// haystack may be news or olds etc as needed
func indexOf(needle PropertyValue, haystack []PropertyValue) (int, bool) {
    for i, x := range haystack {
        if Equal(ToTF(x), ToTF(needle)) {
             return i, true
        }
    }
    return 0, false
}
```

With this algo I think this code can complete what's needed even without having a concept of hash.

Reply:
We can probably postpone refactoring this out until PF implementation.

Reply:
I believe PF also hashes elements but does not allow one to override the has function, unlike SDKv2. Thanks for the code snippets, might be useful once we get to PF but I'd also say postponing is best here.